Fix Windows WYSIWYG linking issue (switching to splitview and resolving links in WYSIWYG)

VasuBhog opened the pull request:
<!-- Thank you for submitting a Pull Request. Please:
* Read our Pull Request guidelines:
  https://github.com/Microsoft/azuredatastudio/wiki/How-to-Contribute#pull-requests.
* Associate an issue with the Pull Request.
* Ensure that the code is up-to-date with the `main` branch.
* Include a description of the proposed changes and how to test them.
-->

This PR addresses #15942. PR is a targeted fix for Windows WYSIWYG issues, as the issue was not reproable on Mac. 

This PR fixes two distinct issues: 
1. Relative Path is incorrectly formatted (fixed using a regex to edit the 
- User inserts an absolute link in WYSIWYG
- exits the cell 
- goes back into the cell and edits the cell (insert a new link)
- switch to split view the link is incorrectly formatted (.\\....\\relativePath) should be (..\\..\\relativePath) - fix shown below.

![parentFolder](https://user-images.githubusercontent.com/23587151/125543932-572d1f20-9eea-4404-8600-c547c3e18c8e.gif)

- Another issue fixed in this PR is that the link resolves incorrectly after exiting the WYSIWYG cell and therefore returns an error message below upon clicking the link: 
![image](https://user-images.githubusercontent.com/23587151/125681360-c5eb73c7-675d-4a2a-b086-49a8ba616d07.png)

I found instead of joining the paths (base notebook folder and target href) incorrectly together we should be resolving it to get the absolute path. This would then fix the entire working scenario so that linking works no matter if WYSIWYG cell gets updated.

barbaravaldez commented:
Tested on Mac and it works as expected. I'll test on Windows as well

chlafreniere commented on src/sql/workbench/contrib/notebook/test/browser/notebookMarkdown.test.ts:
```diff
@@ -56,7 +57,9 @@ suite('NotebookMarkdownRenderer', () => {
-			assert.strictEqual(result.innerHTML, `<p><a href="\\maddy\\test\\.build\\someimageurl" data-href="\\maddy\\test\\.build\\someimageurl" title="\\maddy\\test\\.build\\someimageurl">Link to relative path</a></p>`);
+			let fullPath = path.resolve('.');
+			let diskDrive = fullPath.substring(0, fullPath.indexOf(':') + 1);
+			assert.strictEqual(result.innerHTML, `<p><a href="${diskDrive}\\maddy\\test\\.build\\someimageurl" data-href="${diskDrive}\\maddy\\test\\.build\\someimageurl" title="${diskDrive}\\maddy\\test\\.build\\someimageurl">Link to relative path</a></p>`);
```
Are we breaking network paths?

Charles-Gagnon replied:
nit: while you're here - I would suggest just using some generic term such as foo instead of our names.

VasuBhog replied:
No UNC paths work properly for linking absolute and relative paths via WYSIWYG.

VasuBhog replied:
Since this test does not have to do with unc paths reverted to previous but added test regarding the marked js issue.

Charles-Gagnon commented on src/sql/workbench/contrib/notebook/browser/htmlMarkdownConverter.ts:
```diff
@@ -312,6 +312,8 @@ export function findPathRelativeToContent(notebookFolder: string, contentPath: U
 			relativePath = relativePath.replace(/\s/g, '%20');
 			if (relativePath.startsWith(path.join('..', path.sep) || path.join('.', path.sep))) {
 				return relativePath;
+			} else if (relativePath.match(/^(?!.*\.\\])\.\W/)) {
```
Comment here explaining what you're trying to match

Charles-Gagnon commented on src/sql/workbench/contrib/notebook/browser/outputs/notebookMarkdown.ts:
```diff
@@ -240,7 +241,14 @@ export class NotebookMarkdownRenderer {
 		} else if (href.slice(0, 2) === '..') {
-			return path.join(base, href);
+			if (isWindows) {
+				// we need to format invalid href formats (ex. ....\file to ..\..\file)
```
Why is this invalid only on windows?

VasuBhog replied:
The href we get back when testing on mac all seem to properly formatted and that is why we can do a join properly. The href after it gets parsed from marked.js on windows gets back into the format stated in the example and as result will always resolve incorrectly.

Charles-Gagnon replied:
That seems not great....would probably be worth digging into why that's happening since that sounds like a possible bug there and we should make sure it's not going to affect anything else.

VasuBhog replied:
Yeah discussed this with Barbara as well, we both saw that it was coming via the HTML/marked js component that is being sent and therefore was not able to change it there. We plan on doing a bug bash after this release on linking in general as there are many different entry points, this change was targetting specifically WYSIWYG and resolving the link on windows.

chlafreniere commented on src/sql/workbench/contrib/notebook/test/browser/notebookMarkdown.test.ts:
```diff
@@ -53,12 +54,19 @@ suite('NotebookMarkdownRenderer', () => {
 		let result: HTMLElement = notebookMarkdownRenderer.renderMarkdown({ value: `[Link to relative path](../test/.build/someimageurl)`, isTrusted: true });
 		if (process.platform === 'win32') {
-			assert.strictEqual(result.innerHTML, `<p><a href="\\maddy\\test\\.build\\someimageurl" data-href="\\maddy\\test\\.build\\someimageurl" title="\\maddy\\test\\.build\\someimageurl">Link to relative path</a></p>`);
+			let fullPath = path.resolve('.');
```
I'm not super clear what this is doing. Would the second condition only be true if the enlistment was run from a UNC path? 

I think the test should ideally be just testing if we have a UNC path in the notebook, if WYSIWYG mucks with the math unnecessarily or not (i.e. doesn't have anything to do with what the relative path to the notebook is).

chlafreniere replied:
So, if we need to change this test, then I'm a bit worried that we're changing the implementation of how UNC paths are being handled.

VasuBhog commented:
After discussions with Chris, the relative path test case was not testing UNC paths and therefore we do not necessarily need to check for them. They still work for resolving paths with this change. 
Created a new test to illustrate what marked js gives us back given a relative path which is the reason we need the regex. I simplified the regex to ensure all links to n parent directory should work consistently. 

I've tested the same scenarios as Barbara on Windows:

Relative Links 
Absolute Paths
Weblinks
Anchor links to Section in the same file
Notebook located multiple parents away 

All of these while switching between WYSIWYG and Split View, and exiting + updating the cell.

Charles-Gagnon commented on src/sql/workbench/contrib/notebook/browser/outputs/notebookMarkdown.ts:
```diff
@@ -240,7 +241,16 @@ export class NotebookMarkdownRenderer {
 			return base.replace(/(:\/*[^/]*)[\s\S]*/, '$1') + href;
 		} else if (href.slice(0, 2) === '..') {
-			return path.join(base, href);
+			if (isWindows) {
```
Here you're checking if it's windows before doing the replace, but the other place you're just doing it for everything. We should be consistent with whether we check or not

chlafreniere commented on src/sql/workbench/contrib/notebook/browser/htmlMarkdownConverter.ts:
```diff
@@ -7,6 +7,8 @@ import TurndownService = require('turndown');
 import { URI } from 'vs/base/common/uri';
 import * as path from 'vs/base/common/path';
 import * as turndownPluginGfm from 'sql/workbench/contrib/notebook/browser/turndownPluginGfm';
+import { invalidRelativePathRegex, replaceInvalidLinkPath } from 'sql/workbench/contrib/notebook/test/common/utils';
```
We shouldn't be referencing utils from test folders

VasuBhog replied:
moved to notebook/common

Charles-Gagnon commented on src/sql/workbench/contrib/notebook/test/common/utils.ts:
```diff
@@ -0,0 +1,11 @@
+export const invalidRelativePathRegex = /^\.\.+(?=\.\.)/g;
+
+export function replaceInvalidLinkPath(href: string): string {
+	href = href.replace(invalidRelativePathRegex, '..\\');
```
This doesn't seem to handle more than 2 sets of .. currently, is that intended?

i.e. 

`......\\test.ipynb` is turned into `..\..\test.ipynb` instead of `..\..\..\test.ipynb` as I would expect

VasuBhog replied:
Yeah instead of using the regex which ends up matching the entire case of all preceding ..'s. We can still use that for htmlMarkdownConvertor conditional to check, but for parsing and conversion, this seems much simpler and works for any 2 sets of ..'s.

Charles-Gagnon commented on src/sql/workbench/contrib/notebook/browser/outputs/notebookMarkdown.ts:
```diff
@@ -240,7 +242,15 @@ export class NotebookMarkdownRenderer {
+				href = replaceInvalidLinkPath(href);
+				return path.resolve(base, href);
+			} else {
+				return path.join(base + href);
```
This should be `path.join(base, href)` right?

Charles-Gagnon commented on src/sql/workbench/contrib/notebook/browser/htmlMarkdownConverter.ts:
```diff
@@ -312,6 +313,9 @@ export function findPathRelativeToContent(notebookFolder: string, contentPath: U
 			if (relativePath.startsWith(path.join('..', path.sep) || path.join('.', path.sep))) {
 				return relativePath;
+				// if relativePath contains improper directory format due to marked js parsing returning an invalid path (ex. ....\) then we need to replace it to ensure the directories are formatted properly (ex. ..\..\)
+			} else if (isWindows && relativePath.match(invalidRelativePathRegex)) {
```
Doing the full pattern match seems unnecessary, seems like you could just check if it starts with .. and that would be enough given what the first check is for

VasuBhog replied:
If we check for ones that also start with just .. that would potentially capture ones that are valid format (ex. ..\..\). So this check targets only the invalid ones specifically.

Charles-Gagnon replied:
It seems odd then that you don't do this check in the other location. Again - consistency is key here. If we don't care about verifying that in one place then why do we care about it somewhere else?

Regex matching is expensive and more confusing to read so it's good to avoid it whenever we can - and so my question here is do we actually need this or can we get away with something simpler? For example always passing it to the function and having that just not change anything if the path is already valid (defensive programming) or by doing a simpler string check for starting with `...` or something.

VasuBhog replied:
Sure makes more sense to be defensive in this regard and not use regex then. Will change it to string check. Can also do the same condition in notebookMarkdown as well to address consisitency.

Charles-Gagnon commented on src/sql/workbench/contrib/notebook/common/utils.ts:
```diff
@@ -0,0 +1,14 @@
+export const invalidRelativePathRegex = /^\.\.+(?=\.\.)/g;
+
+export function replaceInvalidLinkPath(href: string): string {
+	// Get first slash of link and use that create relative path format (..\) string
```
This should be put as a doc comment on the function

Charles-Gagnon commented on src/sql/workbench/contrib/notebook/common/utils.ts:
```diff
@@ -0,0 +1,14 @@
+export function replaceInvalidLinkPath(href: string): string {
+	// Get first slash of link and use that create relative path format (..\) string
+	// and then concatenate relative path format string to rest of href path after slash
+	let slashIndex = href.indexOf('\\');
```
For something like this I think you should 

1. Make it defensive against paths that don't necessarily start with .... such that it returns the original string
2. Indicate that it's meant for windows only. Maybe even make a check here that just returns the original string if it's not windows?
3. Be clearly documented regardless about what it does/doesn't support so someone looking at this later is clear on what it's supposed to be accomplishing

VasuBhog replied:
I guess one question I have is if we also have the checks here for isWindows and string.startswith then what would be the point for having them in the conditions that call this function?

Charles-Gagnon replied:
That's a good question - and in general I think it would be fine (and easier to understand) to get rid of those checks if the function is able to handle all cases regardless.

VasuBhog replied:
I think the only reason I would not go down that route is that we would have to take in all the cases from notebookMarkdown and htmlMarkdownConvertor (since the case is in an else if condition). I will keep it for how it is now since it is defensive against all paths, as it might be easier to call the function in the future depending on what the cases look like via that entry point.

Charles-Gagnon reviewed (approved):
We should try and follow up with figuring out what the actual issue is here since it seems likely that it would cause problems elsewhere - but this is fine for a targeted fix.

VasuBhog commented on src/sql/workbench/contrib/notebook/common/utils.ts:
```diff
@@ -0,0 +1,26 @@
+ * @returns properly formatted relative path
+ */
+export function replaceInvalidLinkPath(href: string): string {
+	if (isWindows && href.startsWith('...')) {
```
Done logic in one place.
